### PR TITLE
Issue 2

### PR DESCRIPTION
**Pull request by Developer-Assessment**

### CODE REVIEW

File: app/controllers/api/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+class Api::UsersController < ApplicationController
+  protect_from_forgery with: :null_session
+
+  def create
+    user = User.new(user_params)
+    if(user.save)
```
Comment (dgoyal-mdsol) [inline app/controllers/api/users_controller.rb]:
```suggestion
    if user.save
```

File: app/controllers/api/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+      render json: { message: "User Created Successfully", errors: nil}, status: 201
+    else
+      render json: { message: "Failed To Create User", errors: user.errors.messages}, status: 406
+    end
+  end
+
+  private
+  def user_params
```
Comment (dgoyal-mdsol) [inline app/controllers/api/users_controller.rb:15]: add a new line before it.

File: config/routes.rb [hunk 3]
```diff
@@ -1,3 +1,6 @@
 Rails.application.routes.draw do
+  namespace :api do
+    post 'users/create'
```
Comment (dgoyal-mdsol) [inline config/routes.rb:3]: is it working?

File: features/user.feature [hunk 4]
```diff
@@ -0,0 +1,90 @@
+Feature: User Registration
+  Creating a User
+
+  Scenario: I want to create a user with all required parameters
+    When  To create user I send a POST request to "api/users/create" with the following
```
Comment (dgoyal-mdsol) [inline features/user.feature:5]:
```suggestion
    When  I send a POST request to "api/users/create" with the following
```

File: features/user.feature [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    Then I receive following JSON response from create user API
+    """
+      {
+        "message": "User Created Successfully",
+        "errors": null
+      }
+    """
+    And The response status of create user API should be "201"
+
+  Scenario: User cann't be created without name
```
Comment (dgoyal-mdsol) [inline features/user.feature]:
```suggestion
  Scenario: I won't be able to create user without name
```

File: features/user.feature [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        "errors": {
+          "name": [
+            "Name should not be empty"
+          ]
+        }
+      }
+    """
+    And The response status of create user API should be "406"
+
+  Scenario: User cann't be created without email
```
Comment (dgoyal-mdsol) [inline features/user.feature]:
```suggestion
  Scenario: I won't be able to create user without email
```

File: features/user.feature [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        "errors": {
+          "email": [
+            "Email Should not be empty"
+          ]
+        }
+      }
+    """
+    And The response status of create user API should be "406"
+
+  Scenario: User cann't be created without mobile
```
Comment (dgoyal-mdsol) [inline features/user.feature]:
```suggestion
  Scenario: I won't be able to create user without mobile
```

File: features/user.feature [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+        "errors": {
+          "mobile": [
+            "Mobile Should not be empty"
+          ]
+        }
+      }
+    """
+    And The response status of create user API should be "406"
+
+  Scenario: Password should contail minimum 8 characters
```
Comment (dgoyal-mdsol) [inline features/user.feature]:
```suggestion
  Scenario: I won't be able to create user with password of less than 8 characters.
```

File: features/user.feature [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+      {
+        "message": "Failed To Create User",
+        "errors": {
+          "password": [
+            "Your password must contain at least 8 characters"
+          ]
+        }
+      }
+    """
+    And The response status of create user API should be "406"
```
Comment (dgoyal-mdsol) [inline features/user.feature:122]: Please add new line after it.

File: spec/models/user_spec.rb [hunk 10]
```diff
@@ -0,0 +1,11 @@
...
+RSpec.describe User, type: :model do
+  
+  describe "Validations" do
+    it { should validate_presence_of(:name).with_message("Name should not be empty") }
+    it { should validate_presence_of(:email).with_message("Email Should not be empty") }
+    it { should validate_presence_of(:mobile).with_message("Mobile Should not be empty") }
+    it { should validate_length_of(:password).is_at_least(8).with_message("Your password must contain at least 8 characters")}
+  
+  end
+end
```
Comment (dgoyal-mdsol) [inline spec/models/user_spec.rb:11]: Please add new line after it.

File: features/user.feature (hunk 9 above)
Comment (dgoyal-mdsol) [reply features/user.feature:122]: Also, Could we add one more scenario as follows:
`I won't be able to create user without required fields` and pass all required fields as blank then response should give all errors.

**Review by dgoyal-mdsol — CHANGES_REQUESTED** [review]

File: config/routes.rb (hunk 3 above)
Comment (Developer-Assessment) [reply config/routes.rb:3]: It is working Dinesh

File: features/user.feature (hunk 4 above)
Comment (Developer-Assessment) [reply features/user.feature:5]: When  I send a POST request to "api/users/create" with the following

If I use this i am getting conflict with other features. That is why I have added - To create user

File: features/user.feature (hunk 9 above)
Comment (Developer-Assessment) [reply features/user.feature:122]: Added new scenario as you suggested

Comment (Developer-Assessment) [reply features/user.feature:122]: Added new scenario as you suggested

File: features/user.feature (hunk 4 above)
Comment (dgoyal-mdsol) [reply features/user.feature:5]: What was the reason of conflict? Have you checked that?

File: spec/models/user_spec.rb (hunk 10 above)
Comment (dgoyal-mdsol) [reply spec/models/user_spec.rb:11]: Please add new line after it.

File: spec/models/user_spec.rb [hunk 11]
```diff
@@ -0,0 +1,12 @@
...
+RSpec.describe User, type: :model do
+  
+  describe "Validations" do
+    it { should validate_presence_of(:name).with_message("Name should not be empty") }
+    it { should validate_presence_of(:email).with_message("Email Should not be empty") }
+    it { should validate_presence_of(:mobile).with_message("Mobile Should not be empty") }
+    it { should validate_length_of(:password).is_at_least(8).with_message("Your password must contain at least 8 characters")}
+  
+  end
+
```
Comment (dgoyal-mdsol) [inline spec/models/user_spec.rb]: Remove this blank line.

File: spec/requests/api/users_controller_spec.rb [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+        expect(json_response['message']).to eq("Failed To Create User")
+      end
+
+      it "should contain errors" do
+        json_response = JSON.parse(response.body)
+        expect(json_response['errors']).not_to be_nil
+      end
+    end
+  end
+
```
Comment (dgoyal-mdsol) [inline spec/requests/api/users_controller_spec.rb]: Remove this blank line.

File: features/user.feature (hunk 4 above)
Comment (Developer-Assessment) [reply features/user.feature:5]: I have used same text in booking feature. while running tests I am getting "ambiguous match" error.

File: spec/models/user_spec.rb (hunk 11 above)
Comment (Developer-Assessment) [reply spec/models/user_spec.rb]: Removed blank line on line 11

File: spec/requests/api/users_controller_spec.rb (hunk 12 above)
Comment (Developer-Assessment) [reply spec/requests/api/users_controller_spec.rb]: Removed blank line at :113

**Review by dgoyal-mdsol — APPROVED** [review]

**Pull request merged**